rialto_airflow/mais.py

```diff
@@ -50,43 +50,76 @@ def get_token(client_id: str, client_secret: str, base_url: str) -> str:
+    response = oauth_session.get(url)
     response.raise_for_status()  # Raise HTTPError for bad responses
     return response.json()
 
 
+def page_size() -> int:
+    """Returns the page size for the API request."""
+    return 100  # This is the max value from the API.
+
+
 def fetch_orcid_users(
-    access_token: str, base_url: str, path: str, limit: Optional[int] = None
+    mais_client_id: str,
+    mais_client_secret: str,
+    mais_token_url: str,
+    base_url: str,
+    limit: Optional[int] = None,
 ) -> list[ORCIDRecord]:
     """Fetches ORCID user records from the MAIS ORCID API, handling pagination."""
 
     orcid_users: list[ORCIDRecord] = []
-    per_page = 100  # This max value from the API.
-    url = f"{base_url}/orcid/v1{path}"
-    params = {"page": 1, "per_page": per_page}
-
-    while True:
-        logger.info(f"Fetching page {params['page']} from {url}")
-        data = get_response(access_token, url, params=params)
+    url = f"{base_url}/orcid/v1/users?scope=ANY&page_number=1&page_size={page_size()}"  # first page url
+    access_token = get_token(mais_client_id, mais_client_secret, mais_token_url)
+
+    while True:  # paging
+        logger.info(f"Fetching {url}")
+        tries = 0
+        while True:  # retry loop for token expiration
+            try:
+                # Try to get the response
+                data = get_response(access_token, url)
+                break  # break out of the retry loop if we get a successful response
+            except requests.exceptions.HTTPError as e:
+                # if we get a 401 unathorized error from the API call, try to get a new token
+                if e.response.status_code == 401:
+                    tries += 1
+                    logger.warning(
+                        f"Attempt {tries}: Access token expired. Retrying..."
+                    )
+                    access_token = get_token(
+                        mais_client_id, mais_client_secret, mais_token_url
+                    )
+                    if tries >= 3:
+                        logger.error(
+                            "Failed to fetch new access token after multiple attempts."
+                        )
+                        raise e  # break out of the retry loop if we fail three times to get a new token
+                    continue  # Retry the current request with the new token
+                else:
+                    raise e  # Re-raise the exception if it's not a 401 error
 
         results = data.get("results", [])
         orcid_users.extend(results)
 
         if limit and len(orcid_users) >= limit:
             break
 
-        next_link = data.get("links", {}).get("next")
-        if not next_link:
+        self_link = data.get("links", {}).get("self")
+        last_link = data.get("links", {}).get("last")
+        # if we have reached the last page, the self and last links will be the same
+        if self_link == last_link:
```

Comment:
I think this is fine, or it could be simplified by just stopping when there is no next_link?

Reply:
adjusted -- will also go do the same thing the mais_orcid_client gem ... i think they should work the same way so if one breaks, they all break together